### PR TITLE
Filter out constraint tuple class constructors in hasNoTypeClasses

### PR DESCRIPTION
**Pull request by Mikolaj**
A rough sketch.  I don't know what I'm doing. See #74.

### CODE REVIEW

**Comment by nomeata** [comment]
Thanks, decent start, I can clean up it a bit before merging. But I’ll wait for the discussion in #74, to see if all your issues can be resolved.

**Comment by Mikolaj** [comment]
I'm additionally filtering out `heqTyConName` and `eqTyConName`, because both appear in my code, I think. Now inspection-testing no longer complains about `~` classes spotted, so it seems to work. Let me know if that's a bad idea.

**Comment by Mikolaj** [comment]
I've reverted the last change --- it works great with `''(~)` --- I had no idea to try something like that.

**Comment by nomeata** [comment]
Can you do it for the tuples as well? (Mostly curious, it might still be good to do so by default)

**Comment by Mikolaj** [comment]
Any idea how to try? I did try adding `''(%,,,,,,,,%)` to the list and `''$d(%,,,,,,,,%)` and `''d(%,,,,,,,,%)` and each is rejected with "parse error on input ‘,’".

**Comment by nomeata** [comment]
I would have expected the first to work. Nevermind, not that important.

**Comment by Mikolaj** [comment]
I don't think I will have any more actionable input for now. 

Your tool, after patching, helped me manually specialize my code much more, to the point I've discovered a new bug in GHC and/or plugins (https://gitlab.haskell.org/ghc/ghc/-/issues/23866), which makes GHC 9.4 unusable for me. I still have one GHC version that works correctly, GHC 9.2, though it probably doesn't specialize as well  --- I can't tell, because I needed to add as exceptions all classes that constrain existential variables, so I can't easily detect failures to eliminate these dictionaries in non-existential contests (the lack of an explicit `exists` quantifier in Core and in Haskell surface language doesn't help, either):

```
inspect $ hasNoTypeClassesExcept 'revRankedListProd [''GoodScalar, ''KnownNat, ''OS.Shape, ''AstSpan, ''Show, ''Ord, ''Numeric, ''Num, ''RowSum, ''Typeable, ''IfDifferentiable, ''NFData, ''RealFrac, ''Integral, ''Fractional, ''RealFloat, ''Floating]  
```

BTW, I've also just discovered, using your tool, than GHC 9.8.1-alpha1 specializes worse than GHC 9.4, even in its unsound `-fpolymorphic-specialisation` mode (https://gitlab.haskell.org/ghc/ghc/-/issues/23874).

**Comment by Mikolaj** [comment]
So, does it have a chance to get merged? :)

**Comment by Mikolaj** [comment]
Oh, yes, the CI fails, because I cheated for the old GHCs. :)

File: src/Test/Inspection/Core.hs [hunk 1]
```diff
@@ -31,6 +32,7 @@ import GHC.Utils.Misc
 import GHC.Core.DataCon
 import GHC.Core.TyCon (TyCon, isClassTyCon)
 #else
+import Builtin.Types (isCTupleTyConName)  -- only guessing the module name
```
Comment (nomeata) [inline src/Test/Inspection/Core.hs:35]:
```suggestion
import TysWiredIn (isCTupleTyConName)
```

**Comment by Mikolaj** [comment]
Thank you!

**Pull request merged**